File: apps/web/tools/stand-sync.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { StandSync } from '../src/features/shared/standSync.mjs';

function ready() {
  const sync = new StandSync();
  assert.equal(sync.beginWrite(), null);
  assert(sync.settleRead(sync.beginRead(), true));
  return sync;
}
test('stand writes wait for a confirmed initial snapshot', () => {
  const sync = new StandSync();
  assert.equal(sync.beginWrite(), null);
  sync.settleRead(sync.beginRead(), false);
  assert.equal(sync.beginWrite(), null);
  sync.settleRead(sync.beginRead(), true);
  assert(sync.beginWrite());
});
test('a seat write is single-flight and invalidates the earlier poll', () => {
  const sync = ready();
  const old = sync.beginRead();
  const write = sync.beginWrite();
  assert(write);
  assert.equal(sync.beginWrite(), null);
  assert.equal(sync.beginRead(), null);
  assert.equal(sync.settleRead(old, true), false);
  assert(sync.blocked);
});
test('a lost response can reconcile but never replays the seat write', () => {
  const sync = ready();
  const write = sync.beginWrite();
  assert(sync.currentWrite(write));
  const read = sync.beginRead({ reconcile: true });
  assert(sync.settleRead(read, true));
  assert.equal(sync.beginWrite(), null, 'reconciliation still belongs to the current write');
  assert(sync.finishWrite(write));
  assert(!sync.blocked);
  assert(sync.beginWrite());
});
test('failed reconciliation locks actions until a successful fresh read', () => {
  const sync = ready();
  const write = sync.beginWrite();
  sync.settleRead(sync.beginRead({ reconcile: true }), false);
  sync.finishWrite(write);
  assert.equal(sync.beginWrite(), null);
  sync.settleRead(sync.beginRead(), true);
  assert(sync.beginWrite());
});
test('late success cannot hide a newer stand read failure', () => {
  const sync = ready();
  const old = sync.beginRead();
  const latest = sync.beginRead();
  sync.settleRead(latest, false);
  assert.equal(sync.settleRead(old, true), false);
  assert(sync.blocked);
});
test('replacement session cannot accept a previous account snapshot or write', () => {
  let token = 'first';
  const sync = new StandSync(() => token);
  sync.settleRead(sync.beginRead(), true);
  const write = sync.beginWrite();
  const read = sync.beginRead({ reconcile: true });
  token = 'second';
  assert.equal(sync.currentWrite(write), false);
  assert.equal(sync.settleRead(read, true), false);
});
test('unmount and StrictMode remount invalidate old requests', () => {
  const sync = ready();
  const old = sync.beginRead();
  sync.dispose();
  assert.equal(sync.beginRead(), null);
  assert.equal(sync.settleRead(old, true), false);
  sync.activate();
  assert.equal(sync.settleRead(old, true), false);
  assert(sync.settleRead(sync.beginRead(), true));
  assert(sync.beginWrite());
});

File: apps/mobile/smarttaxi_app/test/stand_sync_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:smarttaxi_app/features/shared/stand_sync.dart';

StandSync ready() {
  final sync = StandSync();
  sync.settleRead(sync.beginRead()!, true);
  return sync;
}

void main() {
  test('stand writes require a confirmed initial snapshot', () {
    final sync = StandSync();
    expect(sync.beginWrite(), isNull);
    sync.settleRead(sync.beginRead()!, false);
    expect(sync.beginWrite(), isNull);
    sync.settleRead(sync.beginRead()!, true);
    expect(sync.beginWrite(), isNotNull);
  });
  test('seat write is single-flight and rejects an earlier poll', () {
    final sync = ready();
    final old = sync.beginRead()!;
    expect(sync.beginWrite(), isNotNull);
    expect(sync.beginWrite(), isNull);
    expect(sync.beginRead(), isNull);
    expect(sync.settleRead(old, true), isFalse);
    expect(sync.blocked, isTrue);
  });
  test('lost response reconciles without replaying a write', () {
    final sync = ready();
    final write = sync.beginWrite()!;
    expect(sync.currentWrite(write), isTrue);
    sync.settleRead(sync.beginRead(reconcile: true)!, true);
    expect(sync.beginWrite(), isNull);
    expect(sync.finishWrite(write), isTrue);
    expect(sync.blocked, isFalse);
    expect(sync.beginWrite(), isNotNull);
  });
  test('failed read-back stays locked until fresh confirmation', () {
    final sync = ready();
    final write = sync.beginWrite()!;
    sync.settleRead(sync.beginRead(reconcile: true)!, false);
    sync.finishWrite(write);
    expect(sync.beginWrite(), isNull);
    sync.settleRead(sync.beginRead()!, true);
    expect(sync.beginWrite(), isNotNull);
  });
  test('late successful read cannot hide a newer failure', () {
    final sync = ready();
    final old = sync.beginRead()!;
    sync.settleRead(sync.beginRead()!, false);
    expect(sync.settleRead(old, true), isFalse);
    expect(sync.blocked, isTrue);
  });
  test('disposed screen ignores pending reads and writes', () {
    final sync = ready();
    final write = sync.beginWrite()!;
    final read = sync.beginRead(reconcile: true)!;
    sync.dispose();
    expect(sync.settleRead(read, true), isFalse);
    expect(sync.finishWrite(write), isFalse);
    expect(sync.beginRead(), isNull);
  });
}
